pwndbg/aglib/heap/mallocng.py

```diff
@@ -527,11 +581,465 @@ def slot_size(self):
+
+
+class Mallocng(pwndbg.aglib.heap.heap.MemoryAllocator):
+    """
+    Tracks the allocator state.
+    By leveraging the __malloc_context symbol.
+
+    Import this singleton class like:
+    from pwndbg.aglib.heap.mallocng import mallocng as ng
+
+    and make sure that you have run ng.init_if_needed()
+    before you used the object.
+    """
+
+    def __init__(self):
+        self.finished_init: bool = False
+
+        self.ctx_addr: int = 0
+        self.ctx: Optional[MallocContext] = None
+        self.has_debug_syms: bool = False
+        self.secret: int = 0
+        self.hope: bool = True
+
+    def init_if_needed(self):
+        """
+        We want this class to be a singleton, but also we can't
+        initialize it as soon as pwndbg is loaded.
+
+        Users of the object are responsible for calling this to
+        make sure the object is initialized.
+        """
+        if self.finished_init:
+            return
+
+        self.ctx_addr = 0
+        self.ctx = None
+        self.has_debug_syms = False
+        self.secret = 0
+        self.hope = True
+
+        self.set_ctx_addr()
+
+        if self.ctx_addr and self.hope:
+            self.ctx = MallocContext(self.ctx_addr)
+
+        self.finished_init = True
+
+    def set_ctx_addr(self):
+        """
+        Find where the __malloc_context global symbol is. Try using debug information,
+        but if it isn't available try using a heuristic.
+        """
+        self.ctx_addr = pwndbg.aglib.symbol.lookup_symbol_addr("__malloc_context")
+        if self.ctx_addr is not None:
+            self.has_debug_syms = True
+            self.secret = memory.read_pointer_width(self.ctx_addr)
+            return
+
+        # No debug information :(
+        self.has_debug_syms = False
+
+        # We will find the __malloc_context object by searching memory for
+        # the secret.
+        # https://elixir.bootlin.com/musl/v1.2.5/source/src/malloc/mallocng/malloc.c#L50
+        # Extract the secret first.
+        # https://elixir.bootlin.com/musl/v1.2.5/source/src/malloc/mallocng/glue.h#L49
+        at_random = int(pwndbg.auxv.get()["AT_RANDOM"])
+        self.secret = memory.read_pointer_width(at_random + 8)
+
+        secret_matches = list(
+            pwndbg.search.search(
+                pwndbg.aglib.arch.pack(self.secret), executable=False, writable=True, aligned=8
+            )
+        )
+
+        # There are going to be multiple matches. We don't
+        # want those on the stack (actual AT_RANDOM) or heap
+        # (structures copying the secret). We want it either from the libc.so
+        # mapping (if musl is dynamically linked) or the executable's
+        # mapping (if musl is statically linked).
+        possible: List[Tuple[int, str]] = []
+        for sm in secret_matches:
+            mapping_name = pwndbg.aglib.vmmap.find(sm).objfile
+            if "[stack" in mapping_name or "[heap" in mapping_name:
```

Comment:
Other threads stacks may not be `[stack` iirc and we could in theory skip them?

Iirc we had sth like `stacks.get()` or sth